Scenario for Azure Arc enabled servers based out of Azure Stack HCI (#615)
microsoft/azure_arc: docs/azure_arc_jumpstart/azure_arc_servers/azure_stack_hci/azure_stack_hci_windows/_index.md

---
type: docs
title: "Windows Server Virtual Machine"
linkTitle: "Windows Server Virtual Machine"
weight: 1
description: >
---

> ## Deploy a Windows Server Virtual Machine in Azure Stack HCI and connect it to Azure Arc using Powershell
> 
> The following README will guide you on how to use the provided PowerShell script to deploy a Windows Server Virtual Machine on an [Azure Stack HCI](https://docs.microsoft.com/en-us/azure-stack/hci/overview) cluster and connect it as an Azure Arc enabled server.
> 
> This guide will **not** provide instructions on how to deploy and set up Azure Stack HCI and it assumes you already have it provisioned. If you don't have any Azure Stack HCI cluster created, please have a look at the following [Azure Stack HCI 20H2 Evaluation Guide](https://github.com/Azure/AzureStackHCI-EvalGuide).
> 
> The commands below and the Powershell script described in this guide should be run on the **management computer**.

Reviewer: ```suggestion
## Deploy a Windows Server Virtual Machine in Azure Stack HCI and connect it to Azure Arc using Powershell

The following README will guide you on how to use the provided PowerShell script to deploy a Windows Server Virtual Machine on an [Azure Stack HCI](https://docs.microsoft.com/en-us/azure-stack/hci/overview) cluster and connect it as an Azure Arc enabled server.

This guide will **not** provide instructions on how to deploy and set up Azure Stack HCI and it assumes you already have it provisioned. If you don't have any Azure Stack HCI cluster created, please have a look at the following [Azure Stack HCI 20H2 Evaluation Guide](https://github.com/Azure/AzureStackHCI-EvalGuide).

This guide will not provide instructions on how to deploy and set up Azure Stack HCI and it assumes you already have a configured cluster. The commands described in this guide should be run on the management computer or in a host server in a cluster.
```

Author: @likamrat, thanks for the suggestion. The script was built for running from a management computer only. It's quite difficult to create a script that could be run from the management computer and the host server, because a remote PS session to the host server and its guest VMs is needed

  Get-AzResourceProvider -ListAvailable | Select-Object ProviderNamespace, RegistrationState | Select-String  -Pattern "Microsoft.HybridCompute"
  ```

* Create Azure service principal (SP)

    To be able to complete the scenario and its related automation, an Azure service principal assigned with the “Contributor” role is required. To create it, login to your Azure account using PowerShell and run the below command.

    ```powershell
    Connect-AzAccount
    $sp = New-AzADServicePrincipal -DisplayName "<Unique SP Name>" -Role 'Contributor'
    ```

    For example:

    ```powershell
    $sp = New-AzADServicePrincipal -DisplayName "AzureStackHCI-VM-Jumpstart" -Role 'Contributor'
    ```

    This command will create a variable with a secure string as shown below:

    ```shell
    Secret                : System.Security.SecureString
    ServicePrincipalNames : {XXXXXXXXXXXXXXXXXXXXXXXXXXXX, http://AzureStackHCI-VM-Jumpstart}
    ApplicationId         : XXXXXXXXXXXXXXXXXXXXXXXXXXXX
    ObjectType            : ServicePrincipal
    DisplayName           : AzureStackHCI-VM-Jumpstart
    Id                    : XXXXXXXXXXXXXXXXXXXXXXXXXXXX
    Type                  :
    ```

>     To expose the generated password use this code to export the secret:

Reviewer: ```suggestion
    To expose the generated password use the below code to export the secret:
```

    ```powershell
    $BSTR = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($sp.Secret)
    $UnsecureSecret = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($BSTR)
    ```

    Copy and save the Service Principal ApplicationId and Secret as you will need it for later on in the automation.

    > **Note: It is optional but highly recommended to scope the SP to a specific [Azure subscription and resource group](https://docs.microsoft.com/en-us/powershell/module/az.resources/new-azadserviceprincipal?view=azps-5.4.0)**

* Enable CredSSP in the Host Server

>     The Powershell script leverages [CredSSP](https://docs.microsoft.com/en-us/windows/win32/secauthn/credential-security-support-provider) to delegate credentials from the management server to the target server for remote authentication. It's required to allow the host server to receive credentials from a remote computer.

Reviewer: ```suggestion
    The PowerShell script leverages [CredSSP](https://docs.microsoft.com/en-us/windows/win32/secauthn/credential-security-support-provider) to delegate credentials from the management server to the target server for remote authentication. It's required to allow the host server to receive credentials from a remote computer.
```

    ```powershell
    Enable-WSManCredSSP -Role Server
    ```

    ![Screenshot Enable-WSManCredSSP Output](./01.png)

    Confirm if CredSSP enabled the host server to receive credentials:

    ```powershell
    Get-WSManCredSSP
    ```

    ![Screenshot Enable-WSManCredSSP Output](./02.png)

* Create a Virtual Switch

    > **Note: If you already have a Virtual Switch with Internet access you can skip this step.**

>     It's required to have a [Virtual Switch](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/get-started/create-a-virtual-switch-for-hyper-v-virtual-machines#create-a-virtual-switch-by-using-windows-powershell) created in advance, before running the Powershell script. If not yet created, you can use the following powershell command in your host server:

Reviewer: ```suggestion
    It's required to have a [Virtual Switch](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/get-started/create-a-virtual-switch-for-hyper-v-virtual-machines#create-a-virtual-switch-by-using-windows-powershell) created in advance, before running the PowerShell script. If not yet created, you can use the following PowerShell command in your host server:
```

    ```powershell
    New-VMSwitch -Name ExternalSwitch  -NetAdapterName Internet-Access -AllowManagementOS $true

>     ```

Reviewer: Any specific reason for not having both CredSSP and vSwitch automation as part of the PS script?

Author: The vSwitch creation can be added to the PS script. The CredSSP I don't think so. I'm enabling it in the management computer in the PS script. However, in the host server it needs to be manually done, since I can't do a remote PS session without it.

## Automation Flow

For you to get familiar with the automation and deployment flow, below is an explanation.

1. User is editing the PowerShell script environment variables (1-time edit). These variables values are being used throughout the deployment and Azure Arc onboarding.

2. User is running the PowerShell script to deploy a basic Windows Server Virtual Machine on Azure Stack HCI and onboard onto Azure Arc. Runtime script will:
    * Set local OS variables

>     * Download .vhdx with Windows Server 2019 Datacenter installed

Reviewer: ```suggestion
    * Download vhdx file with Windows Server 2019 Datacenter installed
```

>     * Enable CredSSP on the management VM, for allowing credential delegation to the host server
>     * Create a new Virtual Machine based on the .vhdx
>     * Add Virtual Machine to Server Cluster (optional)
>     * Assign Static IP address to Virtual Machine (optional)
>     * Install Azure Arc agent and onboarding

Reviewer: ```suggestion
    * Enable CredSSP on the management VM, for allowing the credential delegation to the host server
    * Create a new Virtual Machine based on the VHDX file
    * Add Virtual Machine to Server Cluster (optional)
    * Assign Static IP address to Virtual Machine (optional)
    * Install Azure Arc agent and onboarding
```

## Deployment

 > **Note: Once more, please make sure all the steps mentioned below are done in your management server.**

As mentioned, this deployment will leverage this [PowerShell script](https://github.com/microsoft/azure_arc/blob/main/azure_arc_servers_jumpstart/azure_stack_hci/powershell/azstack_hci_vm_deploy.ps1) to deploy all the needed componenets. Before running the script, it's important to edit the script variables for matching the parameters of your environment. We can split it in three sets of variables:

1. Environment variables for optional configurations

    * **DHCPEnabled:** Select _$true_ if DHCP is enabled on your environment, _$false_ if not.

>         * If **DHCPEnabled** = _$false_, please fill the following variables. 

Reviewer: this is a bit confusing explanation

> 1. Environment variables for optional configurations
> 
>     * **DHCPEnabled:** Select _$true_ if DHCP is enabled on your environment, _$false_ if not.
>         * If **DHCPEnabled** = _$false_, please fill the following variables. 
>             * **IPAddress:**  Provide the static IP to assign the to the VM.
>             * **Prefix Lenght:** Provide the subnet lenght to assing to the VM.
>             * **DefaultGateway:** Provide the default gateway to assign to the VM.
>             * **DNSServer:** Provide the DNS Server to assign to the VM.
>         * If **DHCPEnabled** = _$true_, please fill attribute _$null_ to all the variables above.
>     * **ServerClusterEnabled:** Select _$true_ if you have a server cluster created, _$false_ if the not.
>         * If **ServerClusterEnabled** = _$true_, please provide the path to the cluster storage in the following format:
>             * **vmdir:** Disk Letter:\ClusterStorage
>         * If **ServerClusterEnabled** = _$false_, please provide the path to the folder where the VM will be created in the following format:
>             * **vmdir:** Disk Letter:\Folder

Reviewer: ```suggestion
1. Environment variables for optional configurations

    * **DHCPEnabled:** Select _$true_ if DHCP is enabled on your environment, _$false_ if not.
        * If **DHCPEnabled** = _$false_, please fill the following variables. 
            * **IPAddress:**  Provide the static IP address to assign to the VM.
            * **Prefix Lenght:** Provide the subnet length to assign to the VM.
            * **DefaultGateway:** Provide the default gateway to assign to the VM.
            * **DNSServer:** Provide the DNS Server to assign to the VM.
        * If **DHCPEnabled** = _$true_, please fill attribute _$null_ to all the variables above.
    * **ServerClusterEnabled:** Select _$true_ if you have a server cluster created, _$false_ if not.
        * If **ServerClusterEnabled** = _$true_, please provide the path to the cluster storage in the following format:
            * **vmdir:** Disk Letter:\ClusterStorage
        * If **ServerClusterEnabled** = _$false_, please provide the path to the folder where the VM will be created in the following format:
            * **vmdir:** Disk Letter:\Folder
```

>     ![Screenshot showing first set of variables in the Powershell script](./03.png)

Reviewer: ```suggestion
    ![Screenshot showing first set of variables in the PowerShell script](./03.png)
```

    * **DomainName:** 'azshci.local'
    * **VMName:** 'ArcJS-VM1'
    * **VSwitchName:** 'Internet-vSwitch'

3. Environment variables to onboard the VM to Azure Arc

    * **subID:** Provide the subscriptionID
    * **appID:** Provide the Service Principal ApplicationID
    * **secret:** Provide the Service Principal Secret
    * **tID:** Provide the tenantID
    * **rgroup:** Provide the Resource Group Name
    * **location:** Provide the Region

    ![Screenshot showing third set of variables in the Powershell script](./05.png)

    As an example:
    * **subID:** "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX"
    * **appID:** "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX"
    * **secret:** "XXXXXXXXXX"
    * **tID:** "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXX"
    * **rgroup:** "arc-vms-rg"
    * **location:** "West Europe"

* After editing the variables, to run the script open PowerShell as an administrator, navigate to the [script folder](https://github.com/microsoft/azure_arc/tree/main/azure_arc_servers_jumpstart/azure_stack_hci/powershell) and run:

  ```powershell
  .\azstack_hci_vm_deploy.ps1
  ```

  > **Note: The script takes some minutes to deploy, specially during the .VHDX download**  

>     ![Screenshot showing the download of the .VHDX file](./06.png)

Reviewer: ```suggestion
    ![Screenshot showing the download of the VHDX file](./06.png)
```

>   > **Note: You'll be asked to provide the credentials for accessing your host server. If prompted with any other authorization requests (CredSSP, networking related, etc), please answer _Yes_ to all of them.**

Reviewer: I am not sure about this one. Can we simplify with extra automation?

Author: Will do my best

    ![Screenshot showing the CredSSP prompt](./07.png)
    ![Screenshot showing the credentials prompt](./08.png)

>     ![Screenshot showing the network related prompts](./09.png)

Reviewer: Align the "this can take some time" to the suggested new language and update screenshots

* Once the script run has finished, the output will look like the following:

 ![Screenshot showing the installation completed!](./10.png)

* The Windows Virtual Machine created in Azure Stack HCI will be projected as a new Azure Arc enabled Server resource.

>  ![Screenshot showing Arc Enabled Server](./11.png)

Reviewer: It looks like the VM Hostname was not updated in the image. Looks generic and should be aligned with the HCI VM and Arc server resource.
